0005-snarkyjs-ecdsa-sha.md

```diff
@@ -0,0 +1,197 @@
+
+In SnarkyJS, these new primitives will be declared as followed:
+
+```ts
+function buildSha(length: 224 | 256 | 385 | 512, nist: boolean) {
+  return {
+    hash(message: Field[]) {
+      let payload = [0, ...message.map((f) => f.value)];
+      return Snarky.sha.create(payload, nist, length).map(Field);
+    },
+  };
+}
+
+const Sha3_224 = buildSha(224, true);
+const Sha3_256 = buildSha(256, true);
+const Sha3_385 = buildSha(385, true);
+const Sha3_512 = buildSha(512, true);
+const Keccak = buildSha(256, false);
+```
+
+Another alternative to the factory pattern above could be to supply the developer with a single function that takes a range of parameters, so that the developer can choose their flavour of SHA3/Keccak on their own.
+
+```ts
+function SHA3(
+  length: 224 | 256 | 385 | 512,
+  nist?: boolean = true
+): { hash(xs: Field[]) };
+```
+
+Developers can then import these functions into their project via
+
+```ts
+import { Sha3_224, Sha3_256, Sha3_385, Sha3_512, Keccak } from "snarkyjs";
+
+Sha3_224.hash(xs);
+// ..
+```
+
+### Alternative Approach
+
+Another possibility is to combine all hash functions, including Poseidon, under a shared namespace `Hash`. Developers will then be able to use these functions by calling `Hash.[hash_name].hash(xs)`. However, this would not be equivalent to the existing `Poseidon` API.
+
+However, the existing `Poseidon` API could be deprecated in favour of the new `Hash` namespace.
+
+```ts
+const Hash = {
+  Poseidon: {
+    hash: (xs: any) => "digest",
+  },
+
+  // ..
+
+  SHA3_224: {
+    hash: (xs: any) => "digest",
+  },
+
+  Keccak: {
+    hash: (xs: any) => "digest",
+  },
+};
+```
+
+Developers would then be able to simply import the `Hash` namespace into their project and use all available hash functions.
+
+```ts
+import { Hash } from "snarkyjs";
+
+Hash.Poseidon.hash(xs);
+Hash.Keccak.hash(xs);
+// ..
+
+// deprecated in favor of Hash.Poseidon.hash
+Poseidon.hash(xs);
+```
+
+It is important to mention that we have a lot of freedom in the the implementation details of the API, as the gates and gadgets are already implemented in OCaml and just need to exposed and wrapped by a developer friendly API.
+
+One goal of the API is to design an easy to use interface for developers as well as maintaining consistency with already existing primitives (Poseidon in this case, Schnorr in the case of ECDSA).
+The usage of both Keccak/SHA3 and Poseidon is the same: The user passes an array of Field elements into the hash function and the output is returned. One important detail is that Keccak/SHA3 only accepts an array of Field element that each are no larger than 1 byte. We will also provide additional helper functions that allow conversion between Field element arrays and hexadecimal encoded strings.
+
+```ts
+function fieldBytesFromHex(xs: Field[]): string;
+function hexToFieldBytes(hex: string): Field[];
```

Comment:
thanks for calling it out that the Keccak interface expects single-byte field elements.

I think we should follow our usual approach here, and introduce a provable type which encodes the constraint of being no more than a single byte long: `UInt8` or `Byte`

The Keccak / SHA3 interface should therefore be:
```ts
function hash(message: UInt8[]): UInt8[]
// TODO what is the return type? we could return a fixed length tuple!
```

Reply:
Btw @Trivo25, your RFC has the `fieldBytesFromHex` and `hexToFieldBytes` interfaces mixed up

Reply:
I'd like to ask if adding `Field` as an acceptable type in the array is helpful. When `hash` is called, we can check if the field is exactly a byte at runtime. That way, we can be more flexible with our types by using `(Field | UInt8)[]`.

Returning tuples would be neat here! I think you keep all the normal array methods, so the added type safety is helpful!

Reply:
+1 to removing the constraint checks in the gadgets and forcing input to be `UInt8` 👍